controllers/crds/cninode_controller.go

```diff
@@ -57,128 +88,99 @@ func (r *CNINodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+					EC2Wrapper: r.EC2Wrapper,
+					Manager:    cleaner,
+					VPCID:      r.VPCID,
+					Log:        ctrl.Log.WithName("eniCleaner").WithName("node"),
 				}
-				if ok := controllerutil.RemoveFinalizer(cniNode, config.NodeTerminationFinalizer); !ok {
-					r.Log.Info("failed to remove finalizer for CNINode, requeuing the request", "cniNode", cniNode.Name)
-					return ctrl.Result{Requeue: true}, nil
+				// Return err if failed to delete leaked ENIs on node so it can be retried
+				if err := cleaner.DeleteLeakedResources(); err != nil {
+					r.Log.Error(err, "failed to cleanup resources during node termination, request will be requeued")
+					return ctrl.Result{}, err
 				}
-				if err := r.Client.Update(ctx, cniNode); err != nil {
-					// Return err so update can be retried
+				if err = r.FinalizerManager.RemoveFinalizers(ctx, cniNode, config.NodeTerminationFinalizer); err != nil {
+					r.Log.Error(err, "failed to remove finalizer on CNINode, will retry", "cniNode", cniNode.Name, "finalizer", config.NodeTerminationFinalizer)
 					return ctrl.Result{}, err
 				}
+				return ctrl.Result{}, nil
+			} else {
+				r.Log.Error(err, "failed to get the node object in CNINode reconciliation, will retry")
+				// Requeue request so it can be retried
+				return ctrl.Result{}, err
+			}
+		} else {
+			// node exists, do not run the cleanup routine(periodic cleanup routine will anyway delete leaked ENIs), remove the finalizer
+			// to proceed with object deletion, and recreate similar object
 
-				// Re-fetch CNINode to verify it is deleted before re-creating it as we use the same name
-				// TODO: do we need to wait till CNINode deletion here?
-				if err := r.Client.Get(ctx, req.NamespacedName, cniNode); err != nil && errors.IsNotFound(err) {
-					if err = retry.OnError(retry.DefaultBackoff, func(error) bool {
-						if err != nil {
-							return true
-						}
-						return false
-					}, func() error {
-						r.Log.Info("creating CNINode after it has been deleted as node still exists", "cniNode", newCNINode.Name)
-						return r.Client.Create(ctx, newCNINode)
-					}); err != nil {
-						// The CNINode is deleted at this point, we cannot re-create the same object
-						// raise node object to publish warning that CNINode is deleted
-						utils.SendNodeEventWithNodeName(r.K8sAPI, node.Name, utils.CNINodeDeleted,
-							fmt.Sprint("CNINode was deleted and could not be re-created by controller"), v1.EventTypeWarning, r.Log)
-
-					}
-					// Requeue request to ensure the state of the object
-					return ctrl.Result{Requeue: true}, nil
-				}
+			// Create a copy without deletion timestamp for creation
+			newCNINode := &v1alpha1.CNINode{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            cniNode.Name,
+					Namespace:       "",
+					OwnerReferences: cniNode.OwnerReferences,
+					// TODO: should we include finalizers at object creation or let controller patch it on Create/Update event?
```

Comment:
I agree. We should create the CRs with finalizer. We can do this as a follow-up.